Ubeswap Limit Order Protocol

BriungRi opened the pull request:
Notable changes and introductions:
- UbeswapOrderBook: An order book that takes a specifiable makerFee for order placements. Fee and fee recipient are managed by an owner
- OrderBookWtihFee: An order book that takes a specifiable makerFee for order placements. Fee and fee recipient are set by the order broadcaster (generally a frontend)
- OrderRFQMixin: Refactor order hashing so it can be reused
- OrderMixin: 
  - Add `extraInteraction` which allows for taker interaction
  - Flip asset transfer order from (taker->maker maker->taker) to (maker->taker taker->maker). This allows order fulfillers to perform flash fulfillments

Rename 1inch to Ubeswap

Changes are accompanied with tests

BriungRi committed:
Add an on-chain order book

BriungRi committed:
Add OrderBookRFQ

BriungRi committed:
Update deployment script for Celo and Alfajores

BriungRi committed:
Fix lint

BriungRi committed:
Add alfajores deployment

BriungRi committed:
Rename 1inch to Ubeswap

BriungRi committed:
Update alfajores deployment

BriungRi committed:
Add UbeswapOrderBook

BriungRi committed:
Deploy UbeswapOrderBook

BriungRi committed:
Update UbeswapOrderBook -> OrderBookWithFee

BriungRi committed:
Add back UbeswapOrderBook

BriungRi committed:
Add back UbeswapOrderBook

BriungRi committed:
Add `extraInteraction`

`extraInteraction` allows market makers to perform a flash fulfillment. This extra variable tips the stack variable limit, so we also refactor makerAmount, takerAmount, and thresholdAmount and pack it into an `amounts` variable.

BriungRi committed:
Remove MINIMUM_FEE

MINIMUM_FEE is useless for anti-spam because someone can just send the fee to themselves anyways.

BriungRi committed:
Add new Alfajores deployment

BriungRi committed:
Remove 1inch logo from README

BriungRi committed:
Flip fulfillment order

Previous:

taker -> maker
order.interaction
extraInteraction
maker -> taker

Now:

maker -> taker
extraInteraction
taker -> maker
order.interaction

The motivation here is to enable takers to do a flash-fulfillment

BriungRi committed:
Deploy to alfajores again

BriungRi committed:
Log Order and Signature rather than keep them in storage

BriungRi committed:
Add UbeswapOrderBook test back in

BriungRi committed:
Fee updates

 - Use 1_000_000 as denominator so we can support as low as 0.01 bps
 - Add a MAX_FEE of 10 bps to UbeswapOrderBook so that users can't be caught off guard

BriungRi committed:
Deploy on Alfajores

BriungRi committed:
Update README

BriungRi committed:
Fix lint

BriungRi committed:
Self-review changes

BriungRi commented on contracts/OrderRFQBook.sol:
```diff
@@ -0,0 +1,40 @@
+import "./LimitOrderProtocol.sol";
+
+/// @title Internal order book for OrderRFQ
+abstract contract OrderRFQBook is EIP712("Ubeswap Limit Order Protocol", "2") {
```
EIP necessary?

BriungRi commented on contracts/OrderRFQBook.sol:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    // solhint-disable-next-line func-name-mixedcase
+    function DOMAIN_SEPARATOR() external view returns(bytes32) {
```
Necessary?

BriungRi committed:
Peer review #1

BriungRi reviewed (commented):
note: self-reviewed

BriungRi committed:
Deploy to Alfajores

diwu1989 commented on contracts/OrderMixin.sol:
```diff
@@ -141,60 +146,92 @@ abstract contract OrderMixin is
-        uint256 makingAmount,
-        uint256 takingAmount,
-        uint256 thresholdAmount
+        uint256[3] memory amounts
```
I think it's easier to define a `struct` for this `OrderAmounts` vs using array

A struct should also solve the too many variable problem because the struct itself only count as 1 stack var.

BriungRi committed:
Change orderAmounts from array to struct

diwu1989 commented on contracts/OrderMixin.sol:
```diff
@@ -222,47 +261,68 @@ abstract contract OrderMixin is
-                // takingAmount / makingAmount <= thresholdAmount / requestedMakingAmount
-                require(takingAmount * requestedMakingAmount <= thresholdAmount * makingAmount, "LOP: taking amount too high");
+                // orderAmounts.takingAmount / orderAmounts.makingAmount <= orderAmounts.thresholdAmount / requestedMakingAmount
+                require(orderAmounts.takingAmount * requestedMakingAmount <= orderAmounts.thresholdAmount * orderAmounts.makingAmount, "LOP: taking amount too high");
```
alright, so this is where we do the limit order price checks

BriungRi committed:
Deploy on alfajores

BriungRi committed:
Deploy on Celo

BriungRi committed:
Add subsidy functionality to UbeswapOrderBook

BriungRi committed:
Deploy new UbeswapOrderBook to Celo/Alfajores

diwu1989 commented:
Im not sure about tightly coupling the order subsidy with the order book itself.
Having a loose-coupling of a subsidy / reward contract from the orderbook allows us to never have to upgrade the orderbook itself, but plug and play a set of interaction callback contracts.

We can model a new `OrderbookNotificationReceiver`, similar to `InteractiveNotificationReceiver`
and rename `notifyFillOrder` to `notifyOrder`, and pass into it the actual `Order` struct

Example of design:
1. limit order can include a `bytes extraInteraction` which encodes the address of the orderbook notification receiver, and any extra interactive data (as defined by the frontend)
2. when order is placed into the order book, AFTER the maker fee is taken, decode the `extraInteraction` and call into the orderbook notifier's `notifyOrder()`
3. notifier can read the order struct and decide how to send reward, caveat is that the notifier need to whitelist the msg.sender to be the OrderBook and not allow other contracts or EOA to call into it.

BriungRi committed:
Modularize RewardDistributor

BriungRi committed:
Update natspec

BriungRi committed:
Allow rewardRate to exceed 100%

BriungRi committed:
Update whitelist contract events

BriungRi committed:
Deploy to celo/alfajores

diwu1989 commented on contracts/UbeswapOrderBook.sol:
```diff
@@ -21,37 +22,29 @@ contract UbeswapOrderBook is OrderBook, Ownable {
         uint256 _fee,
         address _feeRecipient,
-        IERC20 _subsidyCurrency
+        IOrderRewardDistributor _rewardDistributor
```
I think this is still too specific to the way Ubeswap wants to do rewards

diwu1989 commented on contracts/UbeswapOrderBook.sol:
```diff
@@ -21,37 +22,29 @@ contract UbeswapOrderBook is OrderBook, Ownable {
         fee = _fee;
         feeRecipient = _feeRecipient;
-        subsidyCurrency = _subsidyCurrency;
+        rewardDistributor = _rewardDistributor;
```
we shouldn't have to tell the order book which reward distributor to use, it should be definable by the UI that's submitting the order as part of interaction data to fully de-couple the reward logic from the order book itself

diwu1989 commented on contracts/UbeswapOrderBook.sol:
```diff
@@ -69,37 +62,17 @@ contract UbeswapOrderBook is OrderBook, Ownable {
+        rewardDistributor = _rewardDistributor;
     }
 
     function broadcastOrder(
```
the signature for `broadcastOrder` should be changed to:
```
function broadcastOrder(
  LimitOrderProtocol.Order _order,
  bytes calldata _signature,
  bytes calldata _orderInteraction)
```

`OrderInteraction` then get decoded to [address, OrderNotifier data]

The reward distributor is then notified of the order + the input order notifier data, and reward distributor can do w/e it wants once its notified by a whitelisted OrderBook of a new order

BriungRi committed:
Refactor OrderRewards to be a general notification target

BriungRi committed:
Rename Ubeswap LOP -> LOP

BriungRi committed:
Rename UbeswapOrderBook with OrderBookWithFee

BriungRi committed:
Update package.json

BriungRi committed:
Update broadcastOrder natspec

BriungRi committed:
Update Celo Alfajores deployments

diwu1989 commented on contracts/interfaces/IOrderNotificationReceiver.sol:
```diff
@@ -0,0 +1,12 @@
+interface IOrderNotificationReceiver {
+    function notifyOrderBroadcasted(
+        LimitOrderProtocol.Order memory _order,
+        address _caller
```
I think this one is redundant because the `_order` already has `maker`.

Is there a reason why we need to distinguish between `_caller` and `order.maker`?

diwu1989 replied:
I think this one is redundant because the `_order` already has `maker`.

Is there a reason why we need to distinguish between `_caller` and `order.maker`?

diwu1989 commented on contracts/OrderBookWithFee.sol:
```diff
@@ -0,0 +1,84 @@
+        uint256 _fee,
+        address _feeRecipient
+    ) OrderBook(_limitOrderProtocol) {
+        require(_fee <= MAX_FEE, "UOB: Fee exceeds MAX_FEE");
```
find and replace `UOB`

diwu1989 replied:
find and replace `UOB`

diwu1989 commented on contracts/OrderBookWithFee.sol:
```diff
@@ -0,0 +1,84 @@
+        feeRecipient = _feeRecipient;
+    }
+
+    /// @notice Logs an order/signature pair. Also executes the order placement fee.
```
`Broadcasts a signed order.`
`Also charges the order placement fee.`

diwu1989 replied:
`Also charges the order placement fee.`

diwu1989 commented on contracts/OrderBookWithFee.sol:
```diff
@@ -0,0 +1,84 @@
+                );
+            }
+
+            if (_notificationTarget != address(0)) {
```
this block should be moved outside of the `fee`

order notification is orthogonal to taking the fee (notifier can also read back into the orderbook to see if `fee` and `feeRecipient` is set

diwu1989 replied:
this block should be moved outside of the `fee`

order notification is orthogonal to taking the fee (notifier can also read back into the orderbook to see if `fee` and `feeRecipient` is set

diwu1989 commented on contracts/OrderBookRewardDistributor.sol:
```diff
@@ -0,0 +1,91 @@
+    /// @param _rewardRecipient The address that will receive the rewards
+    function notifyOrderBroadcasted(
+        LimitOrderProtocol.Order memory _order,
+        address _rewardRecipient
```
Why `_rewardRecipient` vs `_order.maker`?

BriungRi replied:
Possible to place an order for someone else, right?

diwu1989 commented on contracts/OrderBookRewardDistributor.sol:
```diff
@@ -0,0 +1,91 @@
+    /// @param _rewardRecipient The address that will receive the rewards
+    function notifyOrderBroadcasted(
+        LimitOrderProtocol.Order memory _order,
+        address _rewardRecipient
```
Why `_rewardRecipient` vs `_order.maker`?

diwu1989 commented on contracts/OrderBookRewardDistributor.sol:
```diff
@@ -0,0 +1,91 @@
+import "./helpers/Whitelistable.sol";
+
+/// @title A permissioned reward distribution module to accompany an OrderBook
+contract OrderBookRewardDistributor is
```
Ha, the cashback contract

diwu1989 replied:
Ha, the cashback contract

diwu1989 commented on contracts/OrderBookRewardDistributor.sol:
```diff
@@ -0,0 +1,91 @@
+        LimitOrderProtocol.Order memory _order,
+        address _rewardRecipient
+    ) public onlyWhitelist {
+        uint256 rewardAmount = _order
```
need to be careful when setting reward rate to account for different makerAsset decimal points

diwu1989 replied:
need to be careful when setting reward rate to account for different makerAsset decimal points

diwu1989 commented on contracts/OrderBookRewardDistributor.sol:
```diff
@@ -0,0 +1,91 @@
+            address(rewardCurrency) != address(0) &&
+            rewardCurrency.balanceOf(address(this)) >= rewardAmount
+        ) {
+            rewardCurrency.safeTransfer(_rewardRecipient, rewardAmount);
```
Emit log for reward?

Could make it easy to build a live-dashboard as users are being rebated for limit orders

diwu1989 commented on contracts/OrderBookRewardDistributor.sol:
```diff
@@ -0,0 +1,91 @@
+            address(rewardCurrency) != address(0) &&
+            rewardCurrency.balanceOf(address(this)) >= rewardAmount
+        ) {
+            rewardCurrency.safeTransfer(_rewardRecipient, rewardAmount);
```
Emit log for reward?

Could make it easy to build a live-dashboard as users are being rebated for limit orders

BriungRi replied:
can already track the rewards by filtering the transfer with `from` as the rewarddistributor contract

diwu1989 commented on contracts/OrderMixin.sol:
```diff
@@ -141,60 +148,92 @@ abstract contract OrderMixin is
+    /// @param target Address that will receive swap funds
+    /// @param extraInteraction Optional interaction. If present, is triggered between asset transfers
+    function fillOrderToWithExtraInteraction(
+        Order memory order,
```
the `order` struct is always immutable inside this file, so everywhere that we have `Order memory order` can be optimized to `Order calldata order` since its unmodified

diwu1989 replied:
the `order` struct is always immutable inside this file, so everywhere that we have `Order memory order` can be optimized to `Order calldata order` since its unmodified

BriungRi committed:
Address audit and comments

 - No reentrancy for OrderBookWithFee#broadcastOrder
 - No submitting the same order twice
 - Return remaining balance if OrderBookRewardDistributor is low on reward tokens
 - Fix comments / docs / error strings

BriungRi committed:
Redeploy to Celo and Alfajores
